sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -757,6 +768,93 @@ Exceptions:
 * Throws synchronously with error code `invalid` if the resulting dependency would
   lead to a cycle. This error is omitted when `property::graph::no_cycle_check` is set.
 
+|
+[source,c++]
+----
+node add_barrier(const std::vector<event> eventList = {});
+----
+
+|Adds a barrier to the graph.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `eventList` - Zero or more events to wait for.
+  These events can be graph-limited events (from graph recording) 
+  or general sycl events (such as events external to the graph).
+  If no event are provided, the barrier waits for all the previous nodes 
+  of the graph to complete.
+
+Returns: The command-group function object node which has been added to the graph.
+
+|
+[source,c++]
+----
+node make_external_event(event& recordedEvent, event& externalEvent, bool updatable = false);
+----
+
+|Turns a graph-limited event into an external event on which other operations 
+can wait (including other graphs).
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `recordedEvent` - Graph-limited event that will be linked to 
+  the external event.
+
+* `externalEvent` - Sycl event to be used as external event.
+
+* `updatable` - Flag indicating that users will update the external event 
+  after the graph finalization. Please, note that setting this flag may degrade 
+  graph execution performance for level-zero backend, as updatable events 
```

Comment:
External events are natively supported by CUDA and HIP backend (I don't know for OpenCL @EwanC ?), and can be updated.
As for level-zero, the current API does not allow us to update events once the command-buffer is created. So, in this case, we have to split the graph and handle these updatable external events on the host. (Hopefully, mutable command-list will help to improve that.)

Reply:
Thanks for clarifying, I think it would be useful to have an aspect for querying whether external events are supported natively.

Reply:
Synchronization outside a command-buffer isn't supported in OpenCL at all, so that backend won't be able to support any of this proposal not just update.

We've got pushback before when trying to add extra aspects as it makes the code less portable, i'd be inclined to leave it out and throw an exception for things that aren't supported natively (documented in spec as a limitation)

Reply:
Thanks for clarifying, so the OpenCL backend as of now wouldn't be able to support this extension at all, is there interest in adding this capability to the OpenCL command buffer extension as well?

I would probably argue that if we make this an optional feature and add an aspect for it this would improve portability because then users can query for the capability to avoid it failing on devices which don't support it. Saying that I think it would be okay to leave it out for the extension, but if we were to progress this to a KHR I think we would need it then.

Reply:
> I would probably argue that if we make this an optional feature and add an aspect for it this would improve portability because then users can query for the capability to avoid it failing on devices which don't support it. Saying that I think it would be okay to leave it out for the extension, but if we were to progress this to a KHR I think we would need it then.

See https://github.com/intel/llvm/pull/12486#discussion_r1500567310 (and https://github.com/intel/llvm/pull/12486#discussion_r1513330185) for more general discussion about the role of aspects in SYCL-Graph. It might be worth commenting there if you have strong feelings about aspects for optional features.

> Thanks for clarifying, so the OpenCL backend as of now wouldn't be able to support this extension at all, is there interest in adding this capability to the OpenCL command buffer extension as well?

There's definitely interest, it was an early feature request https://github.com/KhronosGroup/OpenCL-Docs/issues/793
The only thing is that the timelines for adding things to the command-buffer spec aren't as quick as a vendor extension we can control, but we could start nudging this in the OpenCL side.